### PR TITLE
Document §5.1 trigger-setup preconditions as split Partial parity

### LINKED ISSUE #2868: HERDER §5.1: Trigger setup preconditions parity
Spec adherence row sliced from parent #2819 (which was rolled-up scope; the parent hit the lifetime bounce-cap on PR-level convergence — narrower atomic issues should converge cleanly).

## Audit row
**HERDER_SPEC §5.1** — Trigger setup preconditions.

## Scope
- Re-read the §5.1 row in `crates/herder/SPEC_ADHERENCE.md`.
- Cross-check henyey's trigger setup preconditions against stellar-core (`HerderImpl::setupTriggerNextLedger` and its guards).
- Identify whether the current gap is a true parity miss or an acceptable architecture split (app/herder layering).
- If a parity miss: tighten the gap in `crates/herder/` or `crates/app/`, add focused tests, update `PARITY_STATUS.md`.
- If acceptable architecture: document the divergence in `SPEC_ADHERENCE.md` and `PARITY_STATUS.md`, no code change.

## Out of scope
Anything in the sibling §5.1/§5.2 trigger-time / ctValidityOffset work (already tracked in #2816 and #2702).

## References
- Parent: #2819
- Original audit: `crates/herder/SPEC_ADHERENCE.md`
- stellar-core: `stellar-core/src/herder/HerderImpl.cpp` (`setupTriggerNextLedger`)

### PR DESCRIPTION
**Pull request by tomerweller**
Closes #2868

## Summary

Documents the §5.1 trigger-setup preconditions as a split **Partial** implementation across `App::try_trigger_consensus()` and `Herder::trigger_next_ledger()`. Adds a focused test proving the `current_ledger + 1 < tracking_slot` gate blocks nomination when the node is behind. Updates SPEC_ADHERENCE.md and PARITY_STATUS.md to accurately describe the divergence: henyey uses corrective recovery via `assert_lcl_consistency()` instead of stellar-core's fail-fast equality precondition.

## Plan reference

[Converged Plan comment](https://github.com/stellar-experimental/henyey/issues/2868#issuecomment-4510052509)

## Test plan

- [x] cargo fmt --check
- [x] cargo clippy -p henyey-app -- -D warnings
- [x] `cargo test -p henyey-app test_try_trigger_consensus` — both tests pass

## Existing tests preserved

- [x] `cargo test -p henyey-app test_try_trigger_consensus_skips_while_applying` — passes
- [x] `cargo test -p henyey-app test_try_trigger_consensus_skips_when_lcl_behind_tracking_slot` — passes
- [x] `cargo test -p henyey-herder test_assert_lcl_consistency_recovers_when_lcl_equals_tracking_slot` — passes
- [x] `cargo test -p henyey-herder test_assert_lcl_consistency_recovers_when_lcl_ahead_of_tracking` — passes
- [x] `cargo test -p henyey-herder test_trigger_next_ledger_skips_stale_slot` — passes

## Deviations from plan

None.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** CHANGES_REQUESTED

**Summary:** The PR correctly calls out the behind-tracking and `LCL >= tracking` behavior, but it still overstates §5.1 parity.

<details>
<summary>Full review</summary>

- `setupTriggerNextLedger()` in stellar-core suppresses nomination on **both** sides of tx-set construction when apply is active: it `releaseAssert(!mLedgerManager.isApplying())` at entry (`stellar-core/src/herder/HerderImpl.cpp:1239-1249`) and then, after `addTxSet()` side effects, returns early if `mLedgerManager.isApplying()` before nomination (`HerderImpl.cpp:1579-1585`).
- henyey only has the **pre-entry** applying gate in `App::try_trigger_consensus()` (`crates/app/src/app/consensus.rs`), then `Herder::trigger_next_ledger()` re-checks only `lcl_matches_slot` after `build_nomination_value()` / `process_ready_fetching_envelopes()` (`crates/herder/src/herder.rs:2475-2480`). There is no parity-equivalent post-build `is_applying_ledger()` suppression before `scp.nominate(...)`.
- That means the current implementation can still nominate in a window where stellar-core would skip: a concurrent ledger close can flip the applying flag during tx-set build / envelope-drain, while LCL has not yet advanced enough for the Rust `lcl_matches_slot` re-check to fail. This is a protocol-sequencing divergence in the exact path this PR is documenting.
- Because of that, the updated parity docs should not say the `applying` guard "matches" or is fully preserved by the app/herder split. Concretely, please either:
  - implement the missing post-build/apply-time suppression before nomination, matching `HerderImpl.cpp:1579-1585`, **or**
  - downgrade the documentation so §5.1 explicitly records that henyey still lacks stellar-core's second `isApplying()` check and can nominate during an in-flight close window that upstream suppresses.

</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** Diff is limited to one focused app-side test plus parity-doc updates, and I did not find a correctness, readability, or error-handling issue in the changes.
Existing-test obligations for refactor were satisfied: the preserved tests passed locally and the PR's CI checks completed successfully.

<details>
<summary>Full review</summary>

- Linked issue #2868 is `kind: refactor` per the `## Triage Report` comment, so I checked that existing coverage stayed green rather than applying the bug-fix regression-test-body policy.
- The new test `crates/app/src/app/mod.rs::test_try_trigger_consensus_skips_when_lcl_behind_tracking_slot` is actually present in the diff and is non-vacuous: it asserts both `app.herder.is_tracking()` and `current_ledger + 1 < tracking_slot` before calling `try_trigger_consensus()`, so the unchanged attempt counter specifically exercises the behind-LCL dispatcher gate.
- The preserved-test list from the converged plan is still satisfied. I ran these exact tests locally in an isolated worktree and they passed:
  - `cargo test -p henyey-app test_try_trigger_consensus_skips_while_applying -- --exact`
  - `cargo test -p henyey-app test_try_trigger_consensus_skips_when_lcl_behind_tracking_slot -- --exact`
  - `cargo test -p henyey-herder test_assert_lcl_consistency_recovers_when_lcl_equals_tracking_slot -- --exact`
  - `cargo test -p henyey-herder test_assert_lcl_consistency_recovers_when_lcl_ahead_of_tracking -- --exact`
  - `cargo test -p henyey-herder test_trigger_next_ledger_skips_stale_slot -- --exact`
- CI is green for this PR: the relevant top-level checks, including `Test`, `Doctest`, `Clippy`, `Format`, and the build/test matrix completion jobs, all completed successfully.
- The documentation updates are directionally correct and avoid the earlier overstatement risk: `SPEC_ADHERENCE.md` and `PARITY_STATUS.md` now describe §5.1 as a split partial implementation and explicitly preserve the `LCL >= tracking` divergence via `Herder::assert_lcl_consistency()` rather than claiming full stellar-core parity.
- I did not find a logic bug, missing edge case, or weak error-handling issue introduced by this diff.

</details>

File: crates/app/src/app/mod.rs [hunk 1]
```diff
@@ -4324,6 +4324,51 @@ mod tests {
...
 
+    /// Regression test for #2868 — §5.1 trigger setup preconditions parity.
+    ///
+    /// Verifies that `try_trigger_consensus` skips the consensus trigger when
+    /// the node's LCL is behind the tracking slot (i.e. `current_ledger + 1 <
+    /// tracking_slot`), matching stellar-core's `isSynced()` precondition in
+    /// `setupTriggerNextLedger`.
+    ///
+    /// Parity: stellar-core HerderImpl.cpp — `trackingConsensusLedgerIndex()`
+    /// requires `lastClosedLedger + 1 == trackingConsensusLedgerIndex`.
```
[the remark this answers is not in the document]
Comment (tomerweller) [reply crates/app/src/app/mod.rs]: Addressed in c55f3f3597a3a5a450d77ed7bb1f792478e8a0d8: corrected the comment to reference `nextConsensusLedgerIndex` / `tracking_slot()` instead of `trackingConsensusLedgerIndex`.

File: crates/herder/SPEC_ADHERENCE.md [hunk 2]
```diff
@@ -21,7 +21,7 @@ excluded. SHOULD claims and operational defaults excluded.
 | §4 | State machine transitions | Full | state.rs:82 + herder.rs:980 |
 | §4 | INV: BOOTING regression forbidden | Full | state.rs:82-88 |
 | §4 | trackingConsensusLedgerIndex >= LCL | Full | herder.rs:1066 (corrective; deviation) |
-| §5.1 | Trigger setup preconditions | Partial | herder.rs:2399-2444 (lcl_matches_slot only) |
+| §5.1 | Trigger setup preconditions | Partial | App::try_trigger_consensus (behind/applying) + herder.rs:2399-2444 (lcl_matches_slot); diverges on LCL≥tracking corrective recovery |
```
[the remark this answers is not in the document]
Comment (tomerweller) [reply crates/herder/SPEC_ADHERENCE.md]: Addressed in c55f3f3597a3a5a450d77ed7bb1f792478e8a0d8: added `consensus.rs:121-163` file:line anchor alongside the existing `herder.rs:2399-2444` anchor.

**Review by tomerweller — COMMENTED** [review]
Addressed feedback. Ready for re-review.

Changes in c55f3f35:
- Fixed test comment to reference `nextConsensusLedgerIndex` / `tracking_slot()` (not `trackingConsensusLedgerIndex`)
- Added `consensus.rs:121-163` file:line anchor to SPEC_ADHERENCE.md §5.1 row
- Downgraded parity docs (both SPEC_ADHERENCE.md and PARITY_STATUS.md) to explicitly note henyey lacks stellar-core's post-build `isApplying()` suppression before nomination (HerderImpl.cpp:1579-1585)

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** The diff correctly narrows this to documentation plus a focused coverage addition for the existing split trigger-precondition behavior.
Existing/refactor test obligations are met: the new behind-tracking guard is exercised locally, and the documented corrective-recovery divergence is already covered by herder tests.

<details>
<summary>Full review</summary>

- Reviewed PR #2874 at head `c55f3f3597a3a5a450d77ed7bb1f792478e8a0d8` only.
- Correctness: the docs now describe the implementation split accurately. `App::try_trigger_consensus()` owns the pre-entry `is_tracking` / behind-tracking / applying gates, while `Herder::trigger_next_ledger()` owns the `LCL + 1` stale-slot guard; this matches the current code paths.
- Correctness: the remaining divergence is stated precisely rather than papered over. henyey still uses `assert_lcl_consistency()` corrective recovery when `LCL >= tracking_slot`, whereas stellar-core fail-fast asserts `trackingConsensusLedgerIndex() == LCL`; and stellar-core still has the post-build `isApplying()` suppression that henyey does not.
- Test obligation: issue #2868 triage marks this as **kind: refactor** / **type: enhancement**, so I checked the claimed preserved coverage instead of applying the bug-fix regression policy. The PR body's claimed checks are consistent with that classification.
- Coverage: the new `test_try_trigger_consensus_skips_when_lcl_behind_tracking_slot` exercises the behavior-significant claim added by this PR (`current_ledger + 1 < tracking_slot` skips nomination before incrementing attempts). Existing `assert_lcl_consistency` tests in `henyey-herder` already cover the corrective-recovery side of the documented split.
- Local verification on the PR head succeeded:
  - `cargo fmt --check`
  - `cargo clippy -p henyey-app -- -D warnings`
  - `cargo test -p henyey-app test_try_trigger_consensus -q`
  - `cargo test -p henyey-herder assert_lcl_consistency -q`
- Readability / error handling: no regressions found. The added prose is specific about which guards live in which layer, and the new test is clear about the precondition it is proving.

</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** CHANGES_REQUESTED

**Summary:** The new docs correctly classify §5.1 as split Partial parity, but they still misplace one live divergence and overstate one test comment.
**Summary:** Please tighten the references so the docs match stellar-core's actual guard sequencing exactly.

<details>
<summary>Full review</summary>

- `stellar-core/src/herder/HerderImpl.cpp::setupTriggerNextLedger()` is a fail-fast gate: `releaseAssert(!isApplying())`, `releaseAssert(isTracking())`, `releaseAssert(trackingConsensusLedgerIndex() == lcl.header.ledgerSeq)`, and `releaseAssert(mLedgerManager.isSynced())` before any timer setup. Your updated `SPEC_ADHERENCE.md` §5.1-1 correctly captures the architectural split and the `LCL >= tracking` corrective-recovery divergence in henyey. That part is accurate.
- The new regression test in `crates/app/src/app/mod.rs` overstates parity in its docstring. Henyey's `try_trigger_consensus()` does a soft skip when `current_ledger + 1 < tracking_slot`; that matches the soft early-return in `HerderImpl::triggerNextLedger()` when `!isSynced()`, not the fail-fast `setupTriggerNextLedger()` assert. Please retarget that doc comment to `triggerNextLedger()` and explicitly note that `setupTriggerNextLedger()` is stricter/fatal upstream.
- The `isApplying()` line reference in the new app-layer commentary is off. `HerderImpl.cpp:1440-1447` is inside `setInSyncAndTriggerNextLedger()`'s timer-seq guard; the relevant `isApplying()` skip is in `HerderImpl::triggerNextLedger()` immediately after the out-of-sync check. Please fix those references in the new test/commentary so the parity breadcrumb points at the actual upstream guard.
- `stellar-core` also re-checks `mLedgerManager.isApplying()` after tx-set construction: `if (ledgerSeqToTrigger != slotIndex || mLedgerManager.isApplying()) return;` just before upgrades/nomination. Henyey's `Herder::trigger_next_ledger()` re-checks slot staleness (`lcl_matches_slot`) but does **not** re-check applying at that post-build point.
- Because that missing post-build apply guard lives in `triggerNextLedger()`, `crates/herder/PARITY_STATUS.md` currently overstates parity by leaving the `triggerNextLedger()` row as `Full` and attaching the divergence only to `setupTriggerNextLedger()`. Please move or duplicate that divergence under the `triggerNextLedger()` mapping and downgrade that row from `Full` accordingly.
- For the same reason, `crates/herder/SPEC_ADHERENCE.md` should mention the missing post-build `isApplying()` half of stellar-core's step-10 re-check in the §5.2 pipeline bullets, not only in the §5.1 summary row. Right now the summary table mentions it, but the detailed §5.2 step list still reads as if step 10 fully matches.
- Requested changes for /do Mode B:
  - Update the new `mod.rs` regression-test docstring to cite `HerderImpl::triggerNextLedger()` for the soft `!isSynced()` skip, and explicitly distinguish it from `setupTriggerNextLedger()`'s fail-fast assert path.
  - Fix the stale `HerderImpl.cpp:1440-1447` references for the `isApplying()` guard to the actual `triggerNextLedger()` guard location.
  - In `PARITY_STATUS.md`, mark `triggerNextLedger()` as partial for the missing post-build `isApplying()` re-check (or otherwise document that divergence on that row rather than only on `setupTriggerNextLedger()`).
  - In `SPEC_ADHERENCE.md` §5.2, amend the step-10 bullet to say henyey matches the post-build stale-slot re-check but still lacks stellar-core's concurrent-apply suppression at that same checkpoint.

</details>

**Comment by tomerweller** [comment]
Addressed feedback in 5cc04bb175911fe836e82c2770d92e6895d89576:

- Fixed stale `HerderImpl.cpp:1579-1585` → `1583-1585` for the post-build `isApplying()` guard reference
- Downgraded `triggerNextLedger()` row in PARITY_STATUS.md from `Full` to `Partial` to reflect missing post-build `isApplying()` re-check
- Amended SPEC_ADHERENCE.md §5.2 step-10 bullet to document that henyey matches the post-build stale-slot re-check but lacks stellar-core's concurrent-apply suppression at that checkpoint

Ready for re-review.

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** CHANGES_REQUESTED

**Summary:** The documentation/test additions themselves look consistent with the diff, but this PR does not satisfy the review-pr policy for a `refactor` issue yet. The required PR-body section for preserved existing coverage is missing, and the corresponding CI evidence is not all green at review time.

<details>
<summary>Full review</summary>

- The linked issue triage report classifies #2868 as **Kind: refactor**. Under the review-pr policy, refactor/docs/test-only PRs must include an **"Existing tests preserved"** section in the PR body, and every item listed there must be green in CI before approval. This PR body has `Summary`, `Plan reference`, `Test plan`, and `Deviations from plan`, but it is missing the required **Existing tests preserved** section entirely. Please add that section and list the preserved existing coverage explicitly.
- At review time, the relevant CI picture is still incomplete: `CI/Test`, `CI/Doctest`, and `Quickstart/build / 2 build / 4 build (core, ...)` are still pending. For a refactor/doc-oriented PR, approval has to wait until the preserved-existing-test claims are both present in the body and fully green in CI.
- Diff inspection: the added regression test in `crates/app/src/app/mod.rs` matches the documented `current_ledger + 1 < tracking_slot` skip path in `App::try_trigger_consensus()`, and the docs updates in `SPEC_ADHERENCE.md` / `PARITY_STATUS.md` are internally consistent with that behavior. I did not find a separate correctness or error-handling issue in the changed lines beyond the missing policy-required PR metadata / green-CI evidence above.

</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** APPROVE

**Summary:** Documentation now correctly marks §5.1 as split Partial parity. The remaining behavioral gaps vs stellar-core are accurately identified and scoped.

<details>
<summary>Full review</summary>

- Cross-checked `stellar-core/src/herder/HerderImpl.cpp`:
  - `setupTriggerNextLedger()` enforces `!isApplying()`, `isTracking()`, `trackingConsensusLedgerIndex() == LCL`, and `isSynced()` via `releaseAssert` (`1237-1249`).
  - `triggerNextLedger()` separately soft-skips when out of sync / applying (`1456-1470`) and re-checks `mLedgerManager.isApplying()` after tx-set construction before nomination (`1583-1585`).
- Cross-checked henyey:
  - `App::try_trigger_consensus()` in `crates/app/src/app/consensus.rs` supplies the pre-entry behind/applying/not-tracking gating.
  - `Herder::assert_lcl_consistency()` in `crates/herder/src/herder.rs` repairs `LCL >= tracking` instead of failing fast.
  - `Herder::trigger_next_ledger()` re-checks slot staleness post-build, but not `is_applying`.
- The doc updates in `crates/herder/SPEC_ADHERENCE.md` and `crates/herder/PARITY_STATUS.md` accurately describe both remaining divergences and no longer overstate `setupTriggerNextLedger()` / `triggerNextLedger()` parity.
- The added regression test in `crates/app/src/app/mod.rs` is correctly scoped to the soft “behind => skip” path and does not over-claim coverage of stellar-core’s fail-fast equality invariant.
- No additional parity discrepancy in the touched PR scope stood out from this audit.

</details>

**Comment by tomerweller** [comment]
Addressed feedback. Ready for re-review.

- Added **Existing tests preserved** section to PR body listing all preserved coverage:
  - `test_try_trigger_consensus_skips_while_applying`
  - `test_try_trigger_consensus_skips_when_lcl_behind_tracking_slot`
  - `test_assert_lcl_consistency_recovers_when_lcl_equals_tracking_slot`
  - `test_assert_lcl_consistency_recovers_when_lcl_ahead_of_tracking`
  - `test_trigger_next_ledger_skips_stale_slot`

All listed tests verified passing locally.

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Parity

**Verdict:** APPROVE

**Summary:** Docs now describe §5.1 as split Partial parity without overstating setup/trigger equivalence.
The touched test is scoped to the soft behind-tracking skip, and the remaining stellar-core divergences are called out in the right places.

<details>
<summary>Full review</summary>

- Cross-checked stellar-core `HerderImpl::trackingConsensusLedgerIndex()` (`stellar-core/src/herder/HerderImpl.cpp:165-180`), `setupTriggerNextLedger()` (`1237-1249`), and `triggerNextLedger()` (`1456-1470`, `1583-1585`).
- Cross-checked henyey `App::try_trigger_consensus()` (`crates/app/src/app/consensus.rs:121-163`), `Herder::assert_lcl_consistency()` (`crates/herder/src/herder.rs:1066-1096`), and `Herder::trigger_next_ledger()` (`2399-2481`).
- `crates/herder/SPEC_ADHERENCE.md` now correctly keeps §5.1 trigger setup as **Partial**, citing the real split: app-layer behind/applying/not-tracking suppression plus herder-side stale-slot gating.
- The docs also now correctly preserve both live divergences:
  - `LCL >= tracking_slot` is repaired by `assert_lcl_consistency()` instead of matching stellar-core's fail-fast equality invariant.
  - henyey lacks stellar-core's post-build `isApplying()` suppression before nomination; this is now called out in both the §5.1 summary and the §5.2 step-10 notes, and `crates/herder/PARITY_STATUS.md` downgrades `triggerNextLedger()` accordingly.
- The new app test comment is parity-accurate: it points to the soft `triggerNextLedger()` out-of-sync early return, and explicitly distinguishes that from `setupTriggerNextLedger()`'s fail-fast assert path.
- No additional parity/spec-adherence mismatch stood out in the touched app/herder/doc scope of this PR.
</details>

**Comment by tomerweller** [comment]
## 🔍 Reviewer: Correctness

**Verdict:** APPROVE

**Summary:** Refactor scope stayed tight: one focused app-side test plus documentation that now accurately describes the split Partial parity story.
I verified the issue kind (`refactor`), checked the preserved coverage locally, and found no blocking correctness, readability, or error-handling problems in the diff.

<details>
<summary>Full review</summary>

- Verified issue #2868's triage comment marks this as **Kind: refactor** with the expected obligation to preserve existing behavior / coverage.
- Read the PR body, changed files, commits, and current check context. The diff is limited to `crates/app/src/app/mod.rs`, `crates/herder/SPEC_ADHERENCE.md`, and `crates/herder/PARITY_STATUS.md`.
- The new test in `crates/app/src/app/mod.rs` is well targeted: it explicitly bootstraps tracking, asserts `current_ledger + 1 < tracking_slot`, and then proves `try_trigger_consensus()` leaves `consensus_trigger_attempts` unchanged, so the result is not vacuous.
- The documentation updates now match the actual control flow: app-side `try_trigger_consensus()` handles the behind/applying/not-tracking gates, while herder-side `trigger_next_ledger()` / `assert_lcl_consistency()` still leave the documented Partial divergences around corrective recovery and the missing post-build `isApplying()` re-check.
- I spot-checked the preserved-coverage claim locally with:
  - `cargo test -p henyey-app test_try_trigger_consensus`
  - `cargo test -p henyey-herder test_assert_lcl_consistency_recovers_when_lcl_equals_tracking_slot`
  - `cargo test -p henyey-herder test_assert_lcl_consistency_recovers_when_lcl_ahead_of_tracking`
  - `cargo test -p henyey-herder test_trigger_next_ledger_skips_stale_slot`
  - `cargo clippy -p henyey-app -- -D warnings`
- All of the above passed locally, and I did not find a remaining correctness or test-coverage gap in this refactor-sized change.

</details>
